### PR TITLE
ReflectionObject

### LINKED ISSUE #16: Implement ReflectionObject
It's not supported currently because it requires an instance of an existing class, which means we can't currently implement...

### PR DESCRIPTION
**Pull request by asgrim**
Fixes #16 by implementing `ReflectionObject` which can reflect run-time properties.

### CODE REVIEW

File: src/Reflection/ReflectionClass.php [hunk 1]
```diff
@@ -331,6 +331,18 @@ public function hasProperty($name)
     }
 
     /**
+     * Protected API to add properties at runtime (used in ReflectionObject)
+     *
+     * @param ReflectionProperty $property
+     */
+    protected function addProperty(ReflectionProperty $property)
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Wait... you can add properties at runtime? WADDAFAK? 

If this is allowed, then it should happen in `ReflectionObject` internals, by overriding parent class properties rather than overwriting the `ReflectionClass` internal state.

Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Yes:

```php
$foo = new \stdClass();
$foo->bar = 'baz';
```
If you create a `ReflectionObject` using `$foo` then `bar` is a runtime property.

`ReflectionObject` can't access `ReflectionClass` properties as they're private. However, can modify them to simply append them as we overwrite all the methods anyway.

Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: @asgrim what about having "cachedProperties" (parent class state) and "runtimeProperties" (child class state)?

**Comment by asgrim** [comment]
@Ocramius updated & rebased

File: src/Reflection/ReflectionClass.php [hunk 2]
```diff
@@ -79,12 +79,15 @@ public static function export($className = null)
...
-        $format  = "Class [ <user> class %s%s%s ] {\n";
+        $isObject = $this instanceof ReflectionObject;
+
+        $format  = "%s [ <user> class %s%s%s ] {\n";
         $format .= "  @@ %s %d-%d\n\n";
         $format .= "  - Constants [%d] {%s\n  }\n\n";
         $format .= "  - Static properties [%d] {%s\n  }\n\n";
         $format .= "  - Static methods [%d] {%s\n  }\n\n";
         $format .= "  - Properties [%d] {%s\n  }\n\n";
+        $format .= ($isObject ? "  - Dynamic properties [%d] {%s\n  }\n\n" : '%s%s');
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: The parallelism between line 90 and line [153~154](https://github.com/Roave/BetterReflection/pull/92/files#diff-0b62f5919f325c15e41b6b4c09690ff7R151) makes this quite hard to follow. Maybe just hardcode the values in this particular line?

File: src/Reflection/ReflectionObject.php [hunk 3]
```diff
@@ -0,0 +1,465 @@
+<?php
+
+namespace BetterReflection\Reflection;
+
+use BetterReflection\Reflector\ClassReflector;
+use BetterReflection\SourceLocator\EvaledCodeSourceLocator;
+use PhpParser\Node\Stmt\Property as PropertyNode;
+use PhpParser\Builder\Property as PropertyNodeBuilder;
+
+class ReflectionObject extends ReflectionClass
```
Comment (Ocramius) [inline src/Reflection/ReflectionObject.php]: In order to avoid future breakages (since you designed this class as a decorator, but then used inheritance due to how Reflection works in internals) you need a test that verifies that ALL public methods of this class are declared in this class.

File: test/unit/Fixture/ClassForHinting.php [hunk 4]
```diff
@@ -4,4 +4,5 @@
 
 class ClassForHinting
 {
+    public $someProperty;
```
Comment (Ocramius) [inline test/unit/Fixture/ClassForHinting.php]: This class is likely used somewhere else. Consider using a separate test asset for this particular test scenario

File: src/Reflection/ReflectionClass.php (hunk 2 above)
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Actually, logic should be replaced in the child class with some duplication. Horrible but correct solution :( 

I'm fine with having a `@TODO` instead of that

File: src/Reflection/ReflectionProperty.php [hunk 5]
```diff
@@ -36,7 +41,7 @@ public function __toString()
     {
         return sprintf(
             'Property [%s %s%s $%s ]',
-            $this->isStatic() ? '' : ($this->isDefault() ? ' <default>' : ' <implicit>'),
+            $this->isStatic() ? '' : ($this->isDefault() ? ' <default>' : ' <dynamic>'),
```
Comment (Ocramius) [inline src/Reflection/ReflectionProperty.php]: Where did `<implicit>` go?

File: src/Reflection/ReflectionObject.php (hunk 3 above)
Comment (asgrim) [inline src/Reflection/ReflectionObject.php]: This is already done in [ReflectionObjectTest.php#126](https://github.com/Roave/BetterReflection/pull/92/files#diff-d4ef1e7f3bf562f17cbf82b1a40fa729R126)

File: src/Reflection/ReflectionProperty.php (hunk 5 above)
Comment (asgrim) [inline src/Reflection/ReflectionProperty.php]: Apparently never used o_O not sure where I got that from initially

File: test/unit/Fixture/ClassForHinting.php (hunk 4 above)
Comment (asgrim) [inline test/unit/Fixture/ClassForHinting.php]: Doesn't affect any other tests... why create another file for no reason?

File: src/Reflection/ReflectionClass.php (hunk 2 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Added `@todo`

File: src/Reflection/ReflectionObject.php (hunk 3 above)
Comment (Ocramius) [inline src/Reflection/ReflectionObject.php]: Oh, I missed that D:

I would probably just have done:

```php
$methods = (new ReflectionClass(ReflectionObject::class))->getMethods();
array_walk(
    $methods,
    function (ReflectionMethod $method) {
        if (! $method->isPublic()) {
            return;
        }
        
        $this->assertSame(ReflectionObject::class, $method->getDeclaringClass()->getName());
    }
);
```

Comment (asgrim) [inline src/Reflection/ReflectionObject.php]: Added some good old commenting to describe the intent of the existing tests that cover this off
